gh-115165: Fix Annotated for immutable types

dave-shawley opened the pull request:
This PR widens the list of ignored exceptions when setting the `__orig_class__` attribute on the return value from an annotated callable. Only `AttributeError` was caught yet instances can raise any `Exception` from `__setattr__`.  After this PR, it will catch any `Exception` instead. This change was motivated by the `uuid.UUID.__setattr__` implementation raises `TypeError`.

<!-- gh-issue-number: gh-115165 -->
* Issue: gh-115165
<!-- /gh-issue-number -->

dave-shawley committed:
gh-115165: Fix Annotated for immutable types

The return value from an annotated callable can raise any exception from
__setattr__ for the `__orig_class__` property.

AlexWaygood commented on Lib/test/test_typing.py:
```diff
@@ -8756,6 +8756,16 @@ class B(str): ...
+
+        annotated = Annotated[C, A("A")]
+        self.assertIsInstance(annotated(), C)
+
```
Could you possibly move this test a bit higher up in the file, so that it's next to the other tests that assert that instances of `_AnnotatedAlias` can be instantiated? I'm talking about these tests here:

https://github.com/python/cpython/blob/846fd721d518dda88a7d427ec3d2c03c45d9fa90/Lib/test/test_typing.py#L8471-L8503

You could also add a test for other kinds of generic aliases being instantiated. For example, this would trigger the same bug that you reported in the issue, even though it has nothing to do with `Annotated`:

```pycon
>>> class Foo[T]:
...     def __setattr__(self, attr, value):
...         raise RuntimeError("no")
...
>>> Foo[int]()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\alexw\AppData\Local\Programs\Python\Python312\Lib\typing.py", line 1142, in __call__
    result.__orig_class__ = self
    ^^^^^^^^^^^^^^^^^^^^^
  File "<stdin>", line 3, in __setattr__
RuntimeError: no
```

dave-shawley replied:
addressed in 7289af40d46f34c92a3159de2a357b2a68c5269b and 7fcb3945ddb595eecad5c3fdd2be7af9a348d4a3

AlexWaygood commented on Lib/typing.py:
```diff
@@ -1127,7 +1127,7 @@ def __call__(self, *args, **kwargs):
         try:
             result.__orig_class__ = self
-        except AttributeError:
+        except Exception:
```
A comment here, stating that we're intentionally catching most exceptions, might be good:

```suggestion
        # Some objects raise TypeError (or something even more exotic)
        # if you try to set attributes on them; guard against that here:
        except Exception:
```

dave-shawley replied:
addressed in ff5e88566e30ae887721e55829a8a3547617e510

JelleZijlstra commented on Misc/NEWS.d/next/Library/2024-02-09-07-20-16.gh-issue-115165.yfJLXA.rst:
```diff
@@ -0,0 +1,3 @@
+``typing.Annotated`` ignores exceptions when setting the ``__orig_class__``
```
```suggestion
``typing.Annotated`` now ignores all exceptions when setting the ``__orig_class__``
```
Make it a little more clear what changed

dave-shawley replied:
addressed in 1d7e6e5b1a750793d09e5f8922da573c4a1b5c53

JelleZijlstra reviewed (approved):
Thanks!

I think this is safe to backport to the bugfix versions (3.11 and 3.12).

AlexWaygood commented on Misc/NEWS.d/next/Library/2024-02-09-07-20-16.gh-issue-115165.yfJLXA.rst:
```diff
@@ -0,0 +1,3 @@
+``typing.Annotated`` ignores exceptions when setting the ``__orig_class__``
+attribute on objects returned from a callable. Previously only ``TypeError``
+was ignored.
```
```suggestion
Ignore most exceptions when attempting to set the ``__orig_class__`` attribute
on objects returned when calling a generic alias. Previously only
:exc:`AttributeError` was ignored.
```

Also, feel free to give yourself credit if you like; it's common to do so:

```suggestion
Ignore most exceptions when attempting to set the ``__orig_class__`` attribute
on objects returned when calling a generic alias. Previously only
:exc:`AttributeError` was ignored. Patch by Dave Shawley.
```

dave-shawley replied:
🤦 _Previously only TypeError_ ... good 👀

dave-shawley replied:
addressed in 1d7e6e5b1a750793d09e5f8922da573c4a1b5c53

AlexWaygood reviewed (commented):
Thanks! This is really great for a first contribution :)

A few small points below:

dave-shawley committed:
fixup! gh-115165: Fix Annotated for immutable types

Update news blurb for gh-115165

dave-shawley committed:
fixup! gh-115165: Fix Annotated for immutable types

Add comment justifying `except Exception`

dave-shawley committed:
fixup! gh-115165: Fix Annotated for immutable types

Renamed test_annotated_callable_returning_immutable to
test_instantiate_immutable and moved it near the other instantiation
tests.

dave-shawley committed:
fixup! gh-115165: Fix Annotated for immutable types

Add a test for generic instantiation that trips the same __setattr__
exception in Annotation.

AlexWaygood committed:
Small tweaks

AlexWaygood reviewed (approved):
Thanks!
